tensorflow/core/common_runtime/device_factory.cc

```diff
@@ -63,17 +64,34 @@ int32 DeviceFactory::DevicePriority(const string& device_type) {
 }
 
 // static
-void DeviceFactory::Register(const string& device_type, DeviceFactory* factory,
-                             int priority) {
+string DeviceFactory::SubDeviceType(const string& device_type) {
+  tf_shared_lock l(*get_device_factory_lock());
+  std::unordered_map<string, FactoryItem>& factories = device_factories();
+  auto iter = factories.find(device_type);
+  if (iter != factories.end()) {
+    return iter->second.subdevice_type;
+  }
+
+  // If subdevice type is not found in the device factories, that means
+  // device is not registed. Unit test will direct create kernel without
+  // device registered. In these cases, device type will be used as
+  // subdevice type (framework:op_kernel_test, c:kernel_test)
```

Comment:
Can you remove "Unit test will direct create kernel without device registered." part of the description? I think unit test behavior is less important here. Also remove "(framework:op_kernel_test, c:kernel_test)".

Reply:
done.